driver-core/src/main/com/mongodb/internal/operation/CommandOperationHelper.java

```diff
@@ -159,609 +139,331 @@ public T apply(final T t, final AsyncConnectionSource source, final AsyncConnect
-                                            final FieldNameValidator fieldNameValidator, final Decoder<T> commandResultDecoder,
-                                            final CommandCreator commandCreator, final CommandWriteTransformer<T, R> transformer) {
-        return executeRetryableCommand(binding, database, readPreference, fieldNameValidator, commandResultDecoder, commandCreator,
-                transformer, noOpRetryCommandModifier());
-    }
-
-    static <T, R> R executeRetryableCommand(final WriteBinding binding, final String database, final ReadPreference readPreference,
-                                            final FieldNameValidator fieldNameValidator, final Decoder<T> commandResultDecoder,
-                                            final CommandCreator commandCreator, final CommandWriteTransformer<T, R> transformer,
-                                            final Function<BsonDocument, BsonDocument> retryCommandModifier) {
-        return withReleasableConnection(binding, (source, connection) -> {
-            BsonDocument command = null;
-            MongoException exception;
-            try {
-                command = commandCreator.create(source.getServerDescription(), connection.getDescription());
-                return transformer.apply(connection.command(database, command, fieldNameValidator, readPreference,
-                        commandResultDecoder, binding.getSessionContext(), binding.getServerApi()), connection);
-            } catch (MongoException e) {
-                exception = e;
-                if (!shouldAttemptToRetryWrite(command, e, connection.getDescription().getMaxWireVersion())) {
-                    if (isRetryWritesEnabled(command)) {
-                        logUnableToRetry(command.getFirstKey(), e);
-                    }
-                    throw transformWriteException(exception);
-                }
-            } finally {
-                connection.release();
-            }
-
-            if (binding.getSessionContext().hasActiveTransaction()) {
+        SingleResultCallback<T> addingRetryableLabelCallback = addingRetryableLabelCallback(callback,
+                connection.getDescription().getMaxWireVersion());
+        connection.commandAsync(database, command, new NoOpFieldNameValidator(), primary(), new BsonDocumentCodec(),
+                binding.getSessionContext(), binding.getServerApi(),
+                transformingWriteCallback(transformer, connection, addingRetryableLabelCallback));
+    }
+
+    static <R> Supplier<R> decorateWriteWithRetries(final RetryState retryState, final Supplier<R> writeFunction) {
+        return new RetryingSyncSupplier<>(retryState, CommandOperationHelper::chooseAndMutateRetryableWriteException,
+                CommandOperationHelper::shouldAttemptToRetryWrite, writeFunction);
+    }
+
+    static <R> AsyncCallbackSupplier<R> decorateWriteWithRetries(final RetryState retryState,
```

Comment:
My understanding is that we add the "async" suffix to asynchronous methods, e.g., `executeAsync`, but this method is synchronous. The "async" suffix is used in the method name because async methods do not use `return`/`throw` to produce a result, so making that clear from the method name was considered important enough. But I do not see what is achieved by adding this suffix to a synchronous method?